src/assets/PythonCopyFileWithShutil/python-copy-paste.md

```diff
@@ -0,0 +1,136 @@
+
+```py
+# Import shell utility module
+import shutil
+
+#  Copy the content of source_file.txt to destination_file.txt
+shutil.copyfile('source_file.txt', 'destination_file.txt')
+```
+
+#### 4. shutil.copyfileobj() method
+
+```py
+shutil.copyfileobj(fsrc, fdst, length)
+```
+
+This method copies the contents of the source file object to the destination file object. The source and destination must be file-like objects. It does not preserve metadata and does not return any value. The optional length parameter specifies an integer value for buffer size.
+
+```py
+# Import shell utility module
+import shutil
+
+# Create file objects
+source_file_object = open("../hello.txt", "r")
+destination_file_object = open("CopyFile/copyFalseSrcPath.txt", "w")
+
+# Copy the content of the file object to another file object
+shutil.copyfileobj(source_file_object, destination_file_object)
+```
+
+### Preventing common mistakes:
+
+- The source and destination cannot specify the same path in `shutil.copy()`, `shutil.copy2()` and `shutil.copyfile()`. If the source and destination specify the same file path then the exception "SameFileError" will be raised. This error is not raised in shutil.copyfileobj() as it overwrites the file if it already exists.
+- An error "FileNotFound" will be raised if the source specifies a path to a file that is not present.
+
+### Conclusion
+
+Each method provides different aspects of file operations, providing flexibility and power to meet various testing needs. Python's `shutil` module offers essential tools for efficient file operations in software testing:
+
+- `shutil.copy`: Great for basic file copying, including permissions.
+- `shutil.copy2`: Ideal for copying files along with most of their metadata, including timestamps.
+- `shutil.copyfile`: Best for fast, simple content-copying without metadata.
+- `shutil.copyfileobj`: Enables precise copying between file-like objects.
+
+Some other commonly used utility functions that can be helpful while automating our tests are:
+
+1. `shutil.mkdirs()`: Creates directories recursively. Useful for setting up directory structures needed for testing or organizing files.
+2. `shutil.copytree()`: Recursively copies an entire directory tree. Ideal for duplicating complex directory structures, including all files and sub-directories, to create or backup environments for testing or deployment.
+3. `shutil.rmtree()`: Deletes a directory tree recursively.Essential for cleaning up after tests or removing temporary directories and their contents completely.
+4. `shutil.move()`: Moves a file or directory to another location. Efficiently relocates files or directories within the filesystem, facilitating organization or deployment process.
+5. `shutil.make_archive()`: Creates an archieve(such as AIP or TAR) of a directory and it's contents. Useful for packaging directories for distributions or backup purposes.
```

Comment:
```suggestion
5. `shutil.make_archive()`: Creates an archieve (such as AIP or TAR) of a directory and its contents. Useful for packaging directories for distributions or backup purposes.
```
I think there is a difference in English and Nepali about spaces with punctuation marks like `()`, `:` and similar because I have noticed that many don't add a space before the bracket. In English, there is always a space on the outside of the brackets, but not on the inside. You can also ask the English teacher about this if you want, I'm sure she can explain this better than I.